main.go

```diff
@@ -183,6 +151,67 @@ func main() {
 		log.Tracef("Toolexec final command:    %q\n", proxyCmd.Args())
 		proxy.MustRunCommand(proxyCmd)
 
+		return
+	case "warmup":
+		tmp, err := os.MkdirTemp("", "orchestrion-warmup-*")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to create temporary directory: %v\n", err)
+			os.Exit(1)
+		}
+		defer os.RemoveAll(tmp)
+
+		log.Tracef("Initializing warm-up module in %q\n", tmp)
+		var stderr bytes.Buffer
+		cmd := exec.Command("go", "mod", "init", "orchestrion-warmup")
+		cmd.Dir = tmp
+		cmd.Stderr = &stderr
+		if err := cmd.Run(); err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to initialize temporary module: %v\n", err)
+			if stderr.Len() > 0 {
+				fmt.Fprintf(os.Stderr, "Error output:\n%s\n", stderr.String())
+			}
+			os.Exit(1)
+		}
+
+		log.Tracef("Running 'orchestrion pin' in the temporary module...\n")
+		stderr.Reset()
+		cmd = exec.Command(orchestrionBinPath, "pin")
+		cmd.Dir = tmp
+		cmd.Stderr = &stderr
+		if err := cmd.Run(); err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to install orchestrion in temporary module: %v\n", err)
+			if stderr.Len() > 0 {
+				fmt.Fprintf(os.Stderr, "Error output:\n%s\n", stderr.String())
+			}
+			os.Exit(1)
+		}
+
+		log.Tracef("Running `go build -v <modules>` in the temporary module...\n")
+		buildArgs := make([]string, 0, 2+len(args)+11)
+		buildArgs = append(buildArgs, "go", "build")
+		buildArgs = append(buildArgs, args...)
+		buildArgs = append(buildArgs,
+			// All packages we may be instrumenting, plus the standard library.
+			"github.com/datadog/orchestrion/instrument",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/gin-gonic/gin",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi.v5",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/gofiber/fiber.v2",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/gorilla/mux",
+			"gopkg.in/DataDog/dd-trace-go.v1/contrib/labstack/echo.v4",
```

Comment:
Is it really necessary to build the contrib ahead of time even the ones that customers are not using? a standard go build will build the library they are using even without orchestrion, the only new package it the dd-trace-go contrib, not all it's underlying dependencies.

Side question: is a hardcoded list the best way for us to list all the packages we need to build ?

Reply:
This is `warmup` and it's intended largely to platform teams who are providing "ready-made" Docker images... In this case, you want builds made with the image to be fast, and it's okay/expected for the image build itself to pay the toll.

Regarding the ahrd-coded-ness of the list... We need to `//go:generate` this, yes.